File: ainpc-core-plugin/src/test/java/ro/ainpc/engine/FeaturePackDependencyValidatorTest.java
package ro.ainpc.engine;

import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class FeaturePackDependencyValidatorTest {

    @Test
    void returnsNoMissingDependenciesWhenAllDeclaredPacksExist() {
        List<String> missing = FeaturePackDependencyValidator.missingDependencies(
            List.of("medieval", "social", "ainpc-scenario-medieval"),
            List.of("Medieval", "social")
        );

        assertTrue(missing.isEmpty());
    }

    @Test
    void reportsMissingDependenciesInDeclaredOrder() {
        List<String> missing = FeaturePackDependencyValidator.missingDependencies(
            List.of("medieval"),
            List.of("social", "economy", "social")
        );

        assertEquals(List.of("social", "economy"), missing);
    }

    @Test
    void ignoresBlankDependencyValues() {
        List<String> missing = FeaturePackDependencyValidator.missingDependencies(
            List.of("medieval"),
            List.of("", " ", "medieval")
        );

        assertTrue(missing.isEmpty());
    }

    @Test
    void removesTransitiveUnavailablePacksFromCandidateSet() {
        Set<String> available = FeaturePackDependencyValidator.resolveAvailablePackIds(Map.of(
            "quest-addon", List.of("base-addon"),
            "base-addon", List.of("missing-shared"),
            "medieval", List.of()
        ));

        assertEquals(Set.of("medieval"), available);
    }
}

File: ainpc-core-plugin/src/test/kotlin/ro/ainpc/engine/FeaturePackDependencyValidatorTest.kt
package ro.ainpc.engine

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class FeaturePackDependencyValidatorTest {
    @Test
    fun returnsNoMissingDependenciesWhenAllDeclaredPacksExist() {
        val missing = FeaturePackDependencyValidator.missingDependencies(
            listOf("medieval", "social", "ainpc-scenario-medieval"),
            listOf("Medieval", "social")
        )

        assertTrue(missing.isEmpty())
    }

    @Test
    fun reportsMissingDependenciesInDeclaredOrder() {
        val missing = FeaturePackDependencyValidator.missingDependencies(
            listOf("medieval"),
            listOf("social", "economy", "social")
        )

        assertEquals(listOf("social", "economy"), missing)
    }

    @Test
    fun ignoresBlankDependencyValues() {
        val missing = FeaturePackDependencyValidator.missingDependencies(
            listOf("medieval"),
            listOf("", " ", "medieval")
        )

        assertTrue(missing.isEmpty())
    }

    @Test
    fun removesTransitiveUnavailablePacksFromCandidateSet() {
        val available = FeaturePackDependencyValidator.resolveAvailablePackIds(
            mapOf(
                "quest-addon" to listOf("base-addon"),
                "base-addon" to listOf("missing-shared"),
                "medieval" to listOf()
            )
        )

        assertEquals(setOf("medieval"), available)
    }
}
